File: VB.net/2024/2024-004/2024-004.vb
Imports System.IO

Module XmasWordSearch

    ' Function to read the grid from the input file
    Function ReadGrid(filePath As String) As String()
        Dim lines As New List(Of String)
        Try
            For Each line As String In File.ReadLines(filePath)
                If Not String.IsNullOrWhiteSpace(line) Then
                    lines.Add(line.Trim())
                End If
            Next
        Catch ex As Exception
            Console.WriteLine($"Error reading file: {ex.Message}")
        End Try
        Return lines.ToArray()
    End Function

    ' Function to check if the word exists in a given direction
    Function CheckWord(grid() As String, x As Integer, y As Integer, dx As Integer, dy As Integer, word As String, rows As Integer, cols As Integer) As Boolean
        For i As Integer = 0 To word.Length - 1
            Dim nx = x + i * dx
            Dim ny = y + i * dy
            If nx < 0 OrElse ny < 0 OrElse nx >= rows OrElse ny >= cols OrElse grid(nx)(ny) <> word(i) Then
                Return False
            End If
        Next
        Return True
    End Function

    ' Count occurrences of the word "XMAS" in all directions
    Function CountXMAS(grid() As String) As Integer
        Dim targetWord As String = "XMAS"
        Dim directions(,) As Integer = {
            {0, 1}, {1, 0}, {1, 1}, {1, -1}, {0, -1}, {-1, 0}, {-1, -1}, {-1, 1}
        }
        Dim rows As Integer = grid.Length
        Dim cols As Integer = grid(0).Length
        Dim count As Integer = 0

        For r As Integer = 0 To rows - 1
            For c As Integer = 0 To cols - 1
                For i As Integer = 0 To directions.GetLength(0) - 1
                    Dim dx As Integer = directions(i, 0)
                    Dim dy As Integer = directions(i, 1)
                    If CheckWord(grid, r, c, dx, dy, targetWord, rows, cols) Then
                        count += 1
                    End If
                Next
            Next
        Next
        Return count
    End Function

    ' Count all X-MAS patterns in the grid
    Function CountAllXMASPatterns(grid() As String) As Integer
        Dim rows As Integer = grid.Length
        Dim cols As Integer = grid(0).Length
        Dim count As Integer = 0

        For r As Integer = 1 To rows - 2
            For c As Integer = 1 To cols - 2
                Dim center As Char = grid(r)(c)
                Dim topLeft As Char = grid(r - 1)(c - 1)
                Dim topRight As Char = grid(r - 1)(c + 1)
                Dim bottomLeft As Char = grid(r + 1)(c - 1)
                Dim bottomRight As Char = grid(r + 1)(c + 1)

                If center = "A"c Then
                    ' Pattern 1: M.S
                    If topLeft = "M"c AndAlso topRight = "S"c AndAlso bottomLeft = "M"c AndAlso bottomRight = "S"c Then count += 1
                    ' Pattern 2: S.M
                    If topLeft = "S"c AndAlso topRight = "M"c AndAlso bottomLeft = "S"c AndAlso bottomRight = "M"c Then count += 1
                    ' Pattern 3: M.M
                    If topLeft = "M"c AndAlso topRight = "M"c AndAlso bottomLeft = "S"c AndAlso bottomRight = "S"c Then count += 1
                    ' Pattern 4: S.S
                    If topLeft = "S"c AndAlso topRight = "S"c AndAlso bottomLeft = "M"c AndAlso bottomRight = "M"c Then count += 1
                End If
            Next
        Next
        Return count
    End Function

    Sub Main()
        Dim filePath As String = "/uploads/input.txt"
        Dim grid() As String = ReadGrid(filePath)

        If grid.Length = 0 Then
            Console.WriteLine("Error: Grid is empty or invalid.")
            Return
        End If

        Dim xmasCount As Integer = CountXMAS(grid)
        Console.WriteLine($"Count of XMAS: {xmasCount}")

        Dim xmasPatterns As Integer = CountAllXMASPatterns(grid)
        Console.WriteLine($"Total X-MAS patterns: {xmasPatterns}")
    End Sub

End Module

File: C++-017/2024/2024-004/2024-004.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <stdexcept>

// Function to check if the word exists in a given direction
bool checkWord(const std::vector<std::string> &grid, int x, int y, int dx, int dy, const std::string &word, int rows, int cols) {
    int len = word.size();
    for (int i = 0; i < len; ++i) {
        int nx = x + i * dx;
        int ny = y + i * dy;
        if (nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] != word[i]) {
            return false;
        }
    }
    return true;
}

// Count occurrences of the word "XMAS" in all directions
int countXMAS(const std::vector<std::string> &grid, int rows, int cols) {
    const std::string targetWord = "XMAS";
    int directions[8][2] = {
        {0, 1},   // Right
        {1, 0},   // Down
        {1, 1},   // Diagonal-right-down
        {1, -1},  // Diagonal-left-down
        {0, -1},  // Left
        {-1, 0},  // Up
        {-1, -1}, // Diagonal-left-up
        {-1, 1}   // Diagonal-right-up
    };

    int count = 0;
    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            for (int d = 0; d < 8; ++d) {
                int dx = directions[d][0];
                int dy = directions[d][1];
                if (checkWord(grid, r, c, dx, dy, targetWord, rows, cols)) {
                    count++;
                }
            }
        }
    }
    return count;
}

// Count all X-MAS patterns
int countAllXMASPatterns(const std::vector<std::string> &grid, int rows, int cols) {
    int count = 0;

    for (int r = 1; r < rows - 1; ++r) {
        for (int c = 1; c < cols - 1; ++c) {
            char center = grid[r][c];
            char topLeft = grid[r - 1][c - 1];
            char topRight = grid[r - 1][c + 1];
            char bottomLeft = grid[r + 1][c - 1];
            char bottomRight = grid[r + 1][c + 1];

            if (center == 'A') {
                // Pattern 1: M.S
                if (topLeft == 'M' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'S') count++;
                // Pattern 2: S.M
                if (topLeft == 'S' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'M') count++;
                // Pattern 3: M.M
                if (topLeft == 'M' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'S') count++;
                // Pattern 4: S.S
                if (topLeft == 'S' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'M') count++;
            }
        }
    }
    return count;
}

int main() {
    const std::string filePath = "/uploads/input.txt";

    // Open the file
    std::ifstream file(filePath);
    if (!file.is_open()) {
        std::cerr << "Error: Unable to open file." << std::endl;
        return 1;
    }

    // Read the grid from the file
    std::vector<std::string> grid;
    std::string line;

    while (std::getline(file, line)) {
        grid.push_back(line);
    }
    file.close();

    if (grid.empty()) {
        std::cerr << "Error: Grid is empty." << std::endl;
        return 1;
    }

    int rows = grid.size();
    int cols = grid[0].size();

    // Validate grid dimensions
    for (const auto &row : grid) {
        if (row.size() != cols) {
            throw std::runtime_error("Error: Grid rows have inconsistent lengths.");
        }
    }

    // Count occurrences of "XMAS"
    int xmasCount = countXMAS(grid, rows, cols);
    std::cout << "Count of XMAS: " << xmasCount << std::endl;

    // Count all X-MAS patterns
    int xmasPatterns = countAllXMASPatterns(grid, rows, cols);
    std::cout << "Total X-MAS patterns: " << xmasPatterns << std::endl;

    return 0;
}
